### PR TITLE
Feature Spotlight: Middleware

### PR DESCRIPTION
**Pull request by gkatsev**

### CODE REVIEW

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+   /* ... */
+  }
+});
+```
+
+You can see a [live example here][live] and the [complete code here][code].
+
+
+
+[pra]: https://github.com/videojs-playbackrate-adjuster
```
Comment (OwenEdwards) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Should be https://github.com/videojs/videojs-playbackrate-adjuster

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  github: gkatsev
+tags:
+  - middleware
+  - video.js 6.0
+---
+
+Middleware is one of the cool new features that is coming to Video.js in version 6.0
+
+With middleware, you are now able to interact with and change how the player and the tech talk to each other.
+The tech is Video.js's abstraction from the player. Separating the player API from the playback technology. With techs, we can plug things like a Flash fallback or a Youtube embed into Video.js
```
Comment (OwenEdwards) [inline source/_drafts/Feature-Spotlight-Middleware.md]: "_... from the player. Separating the player API ..._" should be "_... from the player, separating the player API ..._". And I'd suggest "... or a _YouTube_ embed into Video.js _without changing the external API or the look-and-feel of the player_".

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+A lot of Video.js users may be familiar with middleware from projects like Express.
+Video.js middleware isn't that different from those.
+In both cases you register your middleware against a particular route to call down the chain when the route is triggered.
+In Express routes are based on the url paths.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]:
```
In Express, routes are based on the URL path.
```

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+In Video.js, these routes are based on the video mimetype.
+And like Express, there are "star" (`*`) middleware that match all routes.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]:
```
And, like Express, ...
```

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+There are two important pieces to be aware off with middleware:
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: `off` should be `of`

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+1. dynamic source handling,
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: `,` should be removed.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+2. intercepting the player and tech interactions.
+
+With the dynamic source handling, you could load video with a custom type and source and resolve it asynchronously.
+A good example for this is a video catalog system.
+The page can be rendered with a specific catalog id and a special mimetype, like so:
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: `id` should be `ID` and `mimetype` should be `MIME type`

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+Middleware is one of the cool new features that is coming to Video.js in version 6.0
+
+With middleware, you are now able to interact with and change how the player and the tech talk to each other.
+The tech is Video.js's abstraction from the player. Separating the player API from the playback technology. With techs, we can plug things like a Flash fallback or a Youtube embed into Video.js
+
+A lot of Video.js users may be familiar with middleware from projects like Express.
+Video.js middleware isn't that different from those.
+In both cases you register your middleware against a particular route to call down the chain when the route is triggered.
+In Express routes are based on the url paths.
+In Video.js, these routes are based on the video mimetype.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: `mimetype` should be `MIME type`

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+The page can be rendered with a specific catalog id and a special mimetype, like so:
+```html
+<video controls class="video-js">
+  <source src="123" type="video/my-catalog">
+</video>
+```
+Then, you could register a middleware for that route -- the type `video/my-catalog`.
+```js
+// middleware methods get the player instance as an argument
+videojs.use('video/my-catalog', funciton(player) {
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: `funciton` typo

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+  // middleware are expected to return an object with the methods on it.
+  // It can be a plain object or an instance of something.
+  return {
+
+    // setSource allows you to tell videojs whether you're going to be handling the source or not
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: `videojs` should probably be `Video.js` for consistency.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        next(null, {
+          src: body.sourceUrl,
+          type: body.sourceType
+        })
+      });
+    }
+  };
+});
+```
+Then when Video.js initializes, it'll go through and call the middleware that are set for `video/my-catalog`.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Pretty sure `Then` should have a comma after it, but I could be wrong. Or it could be removed.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+Server Side Ad Insertion (SSAI) is a great fit for middleware. It showcases the ability to intercept the play and tech interactions.
+For example, you have both ads and content in your HLS playlist but you want the controls to have a different times for each.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: `to have a` reads like it should be `to display`, I think.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+When the content video is playing you want to display a timeline of 5 minutes. And for ads, you want to display 30 seconds for the pre-roll.
+Yet, currently, the timeline is showing a combined duration of 5 minutes and 30 seconds.
+So, you can add a middleware that knows when the ad is being played and tells the player that the duration is 30 seconds
+and when the content is playing that the duration is 5 minutes.
+```js
+// register a star-middleware because HLS has two mimetypes
+videojs.use('*', function(player) {
```
Comment (mister-ben) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Should adding an HLS middleware to star be encouraged over adding it for both mime types specifically?

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+      });
+    }
+  };
+});
+```
+Then when Video.js initializes, it'll go through and call the middleware that are set for `video/my-catalog`.
+
+Server Side Ad Insertion (SSAI) is a great fit for middleware. It showcases the ability to intercept the play and tech interactions.
+For example, you have both ads and content in your HLS playlist but you want the controls to have a different times for each.
+When the content video is playing you want to display a timeline of 5 minutes. And for ads, you want to display 30 seconds for the pre-roll.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Should have a comma after `playing`. And probably not after `ads`.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+    duration(dur) {
+      return dur / player.playbackRate();
+    },
+
+    /* ... */
+  };
+});
+```
+So, here, we attach a star-middleware because we want to have it applied to all our videos.
+In `setSource`, we also call `next` direct will `null` and the `srcObj` because we can handle any source.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Looks like this should be `directly` instead of `direct`.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+We also set up our `duration` method to take in the duration from the previous middleware and divide it by the playback rate we get from the player.
+
+If you look at the [code][] you can see some other methods next to duration.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Is `[code][]` a mistake?

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+They're there to make sure other methods that rely on timing get updated.
+The two main ones to note is that there's a `currentTime` and a `setCurrentTime`.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: This sentence doesn't really read correctly. I think it's the plural `ones` and singular verb `is`. Maybe: 

> The two methods to notice are `currentTime` and `setCurrentTime`.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+`currentTime` gets called when we want to know what the current time is.
+`setCurrentTime` is called when we're seeking.
+Because the user is seeking in the shifted time, we want to apply our change operation in reverse.
+Instead of dividing it, we want to multiple it.
```
Comment (misteroneill) [inline source/_drafts/Feature-Spotlight-Middleware.md]: `multiply`

**Review by misteroneill — DISMISSED** [review]
Mostly grammar stuff here. One concern about formatting though. This sort of Markdown source:

```
I am a sentence.
This is another sentence.
Well, so am I, so there.
Fine, I'll take my sentences and go home!
```

Renders out with a `<br>` between each sentence, which doesn't lend itself to readable paragraphs:

I am a sentence.
This is another sentence.
Well, so am I, so there.
Fine, I'll take my sentences and go home!

File: source/_drafts/Feature-Spotlight-Middleware.md (hunk 13 above)
Comment (gkatsev) [reply source/_drafts/Feature-Spotlight-Middleware.md]: I don't know which is better!

File: source/_drafts/Feature-Spotlight-Middleware.md (hunk 18 above)
Comment (gkatsev) [reply source/_drafts/Feature-Spotlight-Middleware.md]: Of course! Silly me.

File: source/_drafts/Feature-Spotlight-Middleware.md (hunk 13 above)
Comment (mister-ben) [reply source/_drafts/Feature-Spotlight-Middleware.md]: Maybe `use` could take an array of types?

Comment (gkatsev) [reply source/_drafts/Feature-Spotlight-Middleware.md]: Could be an option.

File: source/_drafts/Feature-Spotlight-Middleware.md (hunk 16 above)
Comment (gkatsev) [reply source/_drafts/Feature-Spotlight-Middleware.md]: Nope, it's shorthand notation for urls. equivalent to `[code][code]`

File: source/_drafts/Feature-Spotlight-Middleware.md (hunk 17 above)
Comment (gkatsev) [reply source/_drafts/Feature-Spotlight-Middleware.md]: Much better.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+In both cases you register your middleware against a particular route to call down the chain when the route is triggered.
+In Express, routes are based on the url paths.
+In Video.js, these routes are based on the video MIME type.
+And, like Express, there are "star" (`*`) middleware that match all routes.
+
+There are two important pieces to be aware of with middleware:
+1. dynamic source handling
+2. intercepting the player and tech interactions.
+
+With the dynamic source handling, you could load video with a custom type and source and resolve it asynchronously.
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: might be good to have a heading here: "A video catalog example"

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+          type: body.sourceType
+        })
+      });
+    }
+  };
+});
+```
+Then, when Video.js initializes, it'll go through and call the middleware that are set for `video/my-catalog`.
+
+Server Side Ad Insertion (SSAI) is a great fit for middleware. It showcases the ability to intercept the play and tech interactions.
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Might be good to make this a heading as well. "Server Side Ad Insertion (SSAI) and `*` middleware"

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+For example, you have both ads and content in your HLS playlist but you want the controls to display different times for each.
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Line 67 & 68 are a bit hard to parse for me. Maybe it could be some thing like this? "For example, you have a 30 second advertisement and 5 minute video listed in an HLS manifest. You want the controls to show the total time for the ad, and the total time for the video separately. Right now the duration of the two will be added together and shown as 5 minutes and 30 seconds."

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+        // We don't want to do anything special for it, so, pass along the source along with a null.
+        next(null, srcObj);
+      }
+    },
+
+    // this method gets called on the tech and then up the middleware chain providing the values as you go along
+    duration(durationFromTech) {
+      if (areWeCurrentlyPlayingAnAd(durationFromTech)) {
+        // since we're now in an ad, return the ad duration
+        return 30;
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: you may want to mention that these values are just for the sake of the example here. A real implementation would probably get the values from an actual HLS manifest.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+      } else {
+        // we're playing back content, so, return that duration
+        return 5 * 60;
+      }
+    }
+  }
+});
+```
+
+A simple but interesting middleware to look at is the [playbackrate-adjuster][pra].
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: A final heading here for this would be good, perhaps: "The `playbackrate-adjuster` an implementation in the wild"

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+The middleware will change the times of the controls depending on the current rate.
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: ~~The~~ This

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+If you're playing back a 20 minute video and change the rate to 2x, the controls will adjust to display 10 minutes.
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Should probably proceed this with: "For instance" or "In example"

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+
+    duration(dur) {
+      return dur / player.playbackRate();
+    },
+
+    /* ... */
+  };
+});
+```
+So, here, we attach a star-middleware because we want to have it applied to all our videos.
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: Maybe this is a redundant edit, but I think it would read better as: ~~to all our videos~~ to any video regardless of the mime type.

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 27]
```diff
@@ -0,0 +1,187 @@
...
+In `setSource`, we also call `next` directly will `null` and the `srcObj` because we can handle any source.
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: ... we also call `next` directly ~~will~~ with `null` and ...

Comment (brandonocasey) [reply source/_drafts/Feature-Spotlight-Middleware.md]: I would also change it from "because we can handle and source", to "because we want to use this middleware with any and all sources."

File: source/_drafts/Feature-Spotlight-Middleware.md [hunk 28]
```diff
@@ -0,0 +1,187 @@
...
+    setTech(newTech) {
+      tech = newTech;
+    }
+
+    /* ... */
+  };
+});
+```
+
+And then, when the `ratechange` event triggers, we tell Video.js that the duration has changed:
```
Comment (brandonocasey) [inline source/_drafts/Feature-Spotlight-Middleware.md]: "and video.js will update the controls accordingly". It may also be good to link to the code that handles this in Video.js

**Review by brandonocasey — CHANGES_REQUESTED** [review]

File: source/_drafts/Feature-Spotlight-Middleware.md (hunk 26 above)
Comment (gkatsev) [reply source/_drafts/Feature-Spotlight-Middleware.md]: That does read better and more clearly.

File: source/_drafts/Feature-Spotlight-Middleware.md (hunk 28 above)
Comment (gkatsev) [reply source/_drafts/Feature-Spotlight-Middleware.md]: I'm not sure linking the code is necessary or worth it.

**Comment by gkatsev** [comment]
Once this is approved, I'm going to change each paragraph to be one line so that it doesn't come out with a bunch of line-breaks inside.

**Review by brandonocasey — APPROVED** [review]
